design/fallbackcert.md

```diff
@@ -0,0 +1,68 @@
+# Fallback Certificate
+
+Status: Draft
+
+Contour supports virtual host based routing over TLS and utilizes SNI which allows multiple fqdn's to be used on the same network endpoint.
+Unfortunately, some requests are sent and do not have the HOST header applied. 
+When this happens, the request fails since the request does not match any routing rules applied to Envoy.
+
+This design doc looks to enable a fallback certificate, such that when a request is received at Envoy, it will still route to the proper set of endpoints even though standard SNI logic isn't applied.
+
+## Goals
+
+- Allow fallback certificate to be defined 
+- Enable the fallback cert only for specific vhosts
+
+## Non Goals
+
+- Allowing multiple default certs to be defined
+
+## Background
+
+Contour provides virtual host based routing, so that the TLS request is routed to the appropriate service, based on the hostname specified in the HTTPS’s `HOST` header.
+As HOST Header is encrypted during TLS handshake, it can’t be used for virtual host based routing unless client sends HTTPS request specifying hostname using the SNI or the request is first decrypted using a default TLS certificate.
+Some users need to support clients which may not be using SNI.
+When an HTTPS request is received, Envoy needs to first decrypt the request using a default TLS certificate and then based on the HOST header, route it to appropriate service.
+As of now, Contour only provides certificate at virtual host level in an IngressRoute and there is no way to define a default TLS certificate in Contour.
+
+## High-Level Design
+
+Contour will add a new argument to `contour serve` named `--fallback-certificate` which references a secret that is namespaced with a name (e.g. `namespace/name`).
```

Comment:
Thinking out loud here ..

- I think that a flag makes sense, but we may want to see whether we can come up with a name that makes it more clear that this is for service, not for contour itself.

- Just to confirm, would this secret would cause the TLS filter match to update?

- WDYT about having this option in the config file too? The config file has a `tls` struct that contains a mon TLS version, maybe it makes sense to configure the certificate there?

Reply:
1. something like `--vhost-fallback-certificate`? Or `--fqdn-fallback-certificate`? Ideas?
2. This would add another filter chain to match. It gets sorted to the bottom so that it is evaluated last. 
3. Yup config is fine. I'm still confused on what we decide is in the file or not. Maybe we need a doc or just a conversation on that so we're all on the same page.

Reply:
> * something like `--vhost-fallback-certificate`? Or `--fqdn-fallback-certificate`? Ideas?

Yeh, some qualifier like that. We don't need to nail it down now, I just wanted to open the discussion :)

> * This would add another filter chain to match. It gets sorted to the bottom so that it is evaluated last.

👍 

> * Yup config is fine. I'm still confused on what we decide is in the file or not. Maybe we need a doc or just a conversation on that so we're all on the same page.

Yeh I don't know either. My experience is that twiddling args in YAML is a pain, so I'm inclined to add them to config unless there's a good reason not to, xref #1266